File: 0790-Domino-and-Tromino-Tiling/0790-Domino-and-Tromino-Tiling.cpp
class Solution {
public:
    int numTilings(int n) {
        if (n == 1) {
            return 1;
        }
        
        const int mod = 1000000007;
        vector<long long> dp(n, 0);
        vector<long long> dp_incomp(n, 0);
        
        dp[0] = 1;
        dp[1] = 2;
        dp_incomp[1] = 2;
        
        for (int i = 2; i < n; i++) {
            dp[i] =( dp[i - 2] + dp[i - 1] + dp_incomp[i - 1]) % mod;
            dp_incomp[i] = (dp[i - 2] * 2 + dp_incomp[i - 1])% mod;
        }
        
        return dp[n - 1] % mod;
    }
};

File: 0790-Domino-and-Tromino-Tiling/0790-Domino-and-Tromino-Tiling.py
class Solution:
    def numTilings(self, n: int) -> int:
        if n == 1:
            return 1
        
        mod = 10 ** 9 + 7
        dp = [0] * n
        dp_incomp = [0] * n
        
        dp[0] = 1
        dp[1] = 2
        dp_incomp[1] = 2
        
        for i in range(2, n):
            dp[i] = (dp[i - 2] + dp[i - 1] + dp_incomp[i - 1]) % mod
            dp_incomp[i] = (dp[i - 2] * 2 + dp_incomp[i - 1]) % mod
        
        return dp[n-1] % mod
